src/txn/txn_ckpt.c

```diff
@@ -659,10 +662,33 @@ __checkpoint_prepare(WT_SESSION_IMPL *session, bool *trackingp, const char *cfg[
     }
 
     /*
-     * Refresh our snapshot here without publishing our shared ids to the world, doing so prevents
-     * us from racing with the stable timestamp moving ahead of current snapshot. i.e. if the stable
-     * timestamp moves after we begin the checkpoint transaction but before we set the checkpoint
-     * timestamp we can end up missing updates in our checkpoint.
+     * Wait for any transactions that are supposed to be stable to finish committing. This prevents
+     * a race where a transaction can begin committing at a time past stable, and another thread
+     * moves stable past that transaction's commit time, and we start checkpointing before it
+     * finishes; we need it in this checkpoint, but it won't be unless we wait for it. If we do,
+     * then this race reduces to the race described in the next comment.
+     *
+     * Note that arguably the proper solution for this race is to not allow stable to advance past a
+     * transaction that hasn't finished committing (or to make it wait instead of us) but that is
+     * not currently feasible.
+     */
+    if (use_timestamp && txn_global->meta_ckpt_timestamp != WT_TS_NONE) {
+        while (__wt_txn_checkpoint_cannot_start(session)) {
+            __wt_sleep(0, 100 * WT_THOUSAND);
+            wait_time++;
```

Comment:
Could you add a statistic in here as well as the verbose message?

Reply:
Done in [9eea8bf](https://github.com/wiredtiger/wiredtiger/pull/7774/commits/9eea8bffa4267bd7b5ab77b621f191bc50057857), but note what I wrote in the commit message.